### DISCUSSION TITLE
Suggestion about CSV import.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#119] [Ideas]
hello,
observation : I have an existing database and I want to switch to the new software. the CSV format is available in export from the old software and the new one imports the format : great

reality : I have not yet managed to import my fields from the old database to the new one (biblioteQ)

explanations : you have made an interface that allows all freedom but behind the database applies huge constraints with the NOT NULL ? (14 or 16)

I have 10000 articles to enter but I have a lot of empty fields that are non-existent in my export database and NOT NULL in your database started with the ISBN.
Yes, the library works with its barcode and the volunteers of 15/20 years ago did not take the time to enter the ISBNs (I understand them: they probably did not have a shower head)

how to do it? 

suggestion : if technically the database can work without NOT NULL : let the freedom ;
with an auto id for the creation of the article and then everyone manages the information they want to enter in it

I have other suggestions about the import interface but I'll save them for other posts

**Comment by textbrowser** [comment]
Did you notice the the Substitute Value column?

![Screenshot_20211109_073210](https://user-images.githubusercontent.com/10701156/140924855-74692b70-a612-4e6c-90cd-65fd38e1f282.png)

**Comment by meteos77** [comment]
first I think I have a vocabulary problem

table name / translation / library notion / 
fields with not null / unique
+ a not too squareCSV file + 

this gives an impossible result 

I make you a table for explanation

**Comment by textbrowser** [comment]
Yes, please do so. Please include a small CSV file that I can ponder over.

**Comment by meteos77** [comment]
analysis file joint
example export table view as CSV
isbn 978225920800-0
Author : Jean DUTOURD 
title : "The grenade and the suppository
edition : PLON

**Comment by meteos77** [comment]
[102_etape1__2021-10-07_PAPURUS_EXPORT_BASE-FOND-PROPRE.csv.tar.gz](https://github.com/textbrowser/biblioteq/files/7506686/102_etape1__2021-10-07_PAPURUS_EXPORT_BASE-FOND-PROPRE.csv.tar.gz)

[Modèle-BiblioteQ-v1.ods](https://github.com/textbrowser/biblioteq/files/7506688/Modele-BiblioteQ-v1.ods)

[test-modif-perso-not-null.sqlite.tar.gz](https://github.com/textbrowser/biblioteq/files/7506695/test-modif-perso-not-null.sqlite.tar.gz)
![Debian11](https://user-images.githubusercontent.com/56431039/140971463-a58fda1f-ee6d-414b-995c-24dd12a15bbf.png)

I finally understood why! it was the settings of my browser

**Comment by meteos77** [comment]
can you tell me what you think these fields should contain? because I don't understand the nuances
1 - id
2 -lccontrolnumber
3 - callnumber
4 - accession_number

**Comment by textbrowser** [comment]
The ID is the ISBN. All item tables have a primary identifier. For books, this field designates the ISBN-10.
The other fields exist because of Z39.50 and the Library of Congress.

**Comment by textbrowser** [comment]
BQ is almost 20 years old (if one considers its predecessor). There are relationships between the database and the interface which I preferred to have been more tiered. Unfortunately, I focused on features instead of internal separation. Features, safety, portability, and openness to an unknown future.

**Comment by meteos77** [comment]
How many users do you count in 20 years? his name was the same?

Well done everyone is happy to have a lot of features!

can someone who compiles review the interface easily
for example I find the fields authors, editors, place of publication too high compared to what they should contain and I have an elevator to go see MARC tag

I saw that I could specify the size but are there any consequences in the engine?

**Reply by textbrowser** [reply]
The number of people is unknown.

Yes, strings are not free. :) Something has to accommodate them.

https://www.postgresql.org/docs/current/limits.html

**Comment by meteos77** [comment]
I finally managed to import my 7000 records into BiblioteQ !

1 - Why does the csv import require an ISBN-13 when the interface for adding 1 book does not require this ? (this would help me because out of the 7000 records I only have 1500 ISBN-13 :-) (of course I unchecked the ISBN's box)

observation : 
- I have 9 variables not reclassified in BiblioteQ
- I have fields in BiblioteQ that I don't really need

what are the fields that are impacted by the use of 
OpenLibrary queries 
SRU queries
Z39.50 queries

because for example binding_type is not at all necessary for us I would prefer to put the date of purchase of the book or other information

**Comment by textbrowser** [comment]
Because it's an import tool. Remember, it expects that you supply complete and correct data. For incomplete fields, it courteously provides substitute values. Let me make the slight adjustment. :)

**Comment by textbrowser** [comment]
Done. :)

https://github.com/textbrowser/biblioteq/commit/63082fbc579d73814daa9a98b830c188bd9c6c4e

**Reply by meteos77** [reply]
you are really extraordinary
thank you very much
I'm finishing my day so I'll let you know when I get back tomorrow
have a good day

**Comment by textbrowser** [comment]
There's a reminder to create a favorites system for the importer. That is, the templates would be customizable. So, if you create a template, you save it and can reuse it later in another BQ session.

**Comment by meteos77** [comment]
my opinion : I think your interface can be improved : the first version was almost perfect

suggestion 1 : generally a csv contains on the first line the name of the fields, having a box in 2bis to know if the 1st line is taken into account would be a plus

suggestion 2 : zone 3 should remain identical so as not to multiply the type 3 zones

if a person in the future asks you for an import of DVDs ? and a 4th for the photographs ...

only the choice zone (popup) "Books | Popup Templates X" should change the popup : 
if the user selects "Template 1 BiblioteQ Book Column Order" it imports books
if the user selects "Template 2 Biblioteq Members Column Order" it imports members
if the user selects "Template 3 Biblioteq DVD Column Order" he imports DVDs
if the user selects "Template 4 Biblioteq Photographs Column Order" it imports photographs

after you provide in the documentation a .csv template file by data type, it is up to the user to fill in the data like my template on the .ods forum
I put in .ods to be able to put the mandatory fields in bold ... the csv does not allow it

As usual: I have ideas but I don't know if in programming it becomes to ask: crossing Everest or a simple hill :-)

**Reply by textbrowser** [reply]
![Screenshot_20211120_101438](https://user-images.githubusercontent.com/10701156/142731618-e55aa8fb-5768-4595-9dde-c51d198febf0.png)
![Screenshot_20211120_101501](https://user-images.githubusercontent.com/10701156/142731623-2d1aab1f-915a-4f1b-a03c-8140d4ca10eb.png)

**Reply by textbrowser** [reply]
In the first image you'll notice that Ignored CSV Row(s) is set to 1. The import will ignore the 1st row of the CSV file. The 1st row is expected to contain the column names.

**Reply by meteos77** [reply]
Super the modification
I always discover new options :-)

maybe put a separator line in the interface between
Ignored CSV row() 
and
the Add Row Delete Row buttons
it would be clearer if there were 2 distinct options

the proof I got caught
or put the option above item 3
it's more natural for me to specify the csv file then the different fields contained in it

**Reply by textbrowser** [reply]
How about this?

![Screenshot_20211120_120205](https://user-images.githubusercontent.com/10701156/142734891-f2579dbb-c82c-4da9-aa1a-1236822939c2.png)

**Reply by meteos77** [reply]
For me now it is OK (and moreover several lines are not importable super option)

after to appreciate the change of design, you need a new eye (I know that it exists so my judgment is now distorted)

**Reply by textbrowser** [reply]
Libre Office. Let's try again.

![Screenshot_20211120_123236](https://user-images.githubusercontent.com/10701156/142735837-d23c51e3-07ce-4262-8617-8a3cc241a912.png)

**Reply by meteos77** [reply]
the most intuitive I've found
![test-bib-import-member](https://user-images.githubusercontent.com/56431039/142736296-936c8305-8409-445b-81aa-43a5406c8e03.png)

with preview of the csv the first 5 lines and assignment 
the number 1 2 3 after assignment changes with the name of the field in the database

**Reply by textbrowser** [reply]
In the Other option, can you provide non-singular characters? For example, a multi-character word.
In CSV Options, only the first row can be ignored.
I will add a preview pane. :)

**Reply by textbrowser** [reply]
Preview pane added. It contains the first 5 data rows.

**Reply by meteos77** [reply]
1 character only 
in BiblioteQ part 4 the text is Ignored CSV Row(s) in plural

the options replace,append,merge could be useful but for later

**Reply by textbrowser** [reply]
Plural? I see both cases.

**Reply by meteos77** [reply]
I did not succeed in making the import? 
I enclose my test file if you can tell me if I have made a mistake ?
[Test_130_Members-for-import.csv](https://github.com/textbrowser/biblioteq/files/7576448/Test_130_Members-for-import.csv)
![error-members-import](https://user-images.githubusercontent.com/56431039/142766631-0f368f3d-d6b4-4c91-9992-0399a481c044.png)

**Reply by textbrowser** [reply]
You're attempting to stroll the baby before it's been born.

**Reply by meteos77** [reply]
nice expression, in fact you made the decor !
in my country the expression would be "Don't sell the bear's skin before you've killed it"

**Reply by textbrowser** [reply]
:)

**Comment by meteos77** [comment]


**Reply by textbrowser** [reply]
I have not committed the work yet. :P

**Comment by textbrowser** [comment]
Patron import completed! Testing required for PQ.

**Reply by meteos77** [reply]
it's cool bravo : it works well for me : I use sqlite

in any case great option that will allow to switch users on new base without worries.

suggestion for future version 2022 : if it doesn't require too much development : 
1 - if the delimiter (box 2) is not the good one can we have a button "update" to remake the CSV preview in case of bad delimiter


2 - to have the csv preview on the whole file with scrollbar on the side (like the fields of the base) because the function 4 makes it possible to ignore the line(s) as much to see them in the same window of the Navigator of Import CSV.

**Reply by textbrowser** [reply]
A refresh? What if the file contains 100,000,000 rows?

**Reply by textbrowser** [reply]
I'll add a refresh. You may think about your question. :)

**Reply by meteos77** [reply]
suggestion : a modest general library option :-)

**Reply by textbrowser** [reply]
100,000,000 rows is modest.

**Reply by meteos77** [reply]
we have 10,000 books maximum and 130 members; that makes all the difference But I don't dispute that managing 100,000,000 is not the same way, hence the pagination... 
on the other hand the option 4 Ignored CSV Row(s) is just to not import the header which would make more than 1 line hence the 5 lines of preview

**Reply by textbrowser** [reply]
Another option. Or perhaps a means of interrupting the preview mechanism. You see, to solve your problem an infinite preview would work. And suppose 11 years from now another person has 111,111,111 rows. They would find the preview unreasonable. :) So, more options (widgets).

**Reply by textbrowser** [reply]
I'm going to be a bit insane. Most people consider software a tool. A means to an end. If the screwdriver I'm using is causing me discomfort, I select another screwdriver (unless I have none other). So, I'm particular about my screwdrivers and so I learn that those who produce that particular tool realize that there are people who are particular about their selection of tools. That is, there is a relationship between the labor and the tool. Some see tools, I see the things that I intend to labor to finish something. Bad tool, unhappy work.

**Reply by meteos77** [reply]
the translator blocks on the word "widgets".

for me the import csv function is OK

I got screwed because my file had ; as separators and not , hence my option request to reload the preview
but if not you have to reopen the file and everything is OK

function filled at 100%.

**Reply by meteos77** [reply]
In the context of volunteering, the tool does not matter, the search for pleasure is more important.

i know of a library that is not computerized because some elderly people are on duty, a large notebook, a pen, and everything works fine. but the number of users is small. if you are lucky enough to pick up a book with a foreign author's name... it takes time.
but their hobby is priceless... especially in time :-)

**Reply by textbrowser** [reply]
If there is the potential for further enjoyment, then evolve. Enjoyment can be ease.

**Comment by textbrowser** [comment]
Interruptible preview is complete. A refresh widget also included. :)

**Reply by textbrowser** [reply]
If one wishes to view 100 or 999,999,999 rows, it's now possible. The operating system may decide to abort the program anyway. Left to the user's discretion.

**Comment by meteos77** [comment]
I recognize that it was stupid to ask for a preview of 7000 lines for my documents but for the 130 members it allows to check that everything is ok before launching the import. it's a question of dosage

I redid the process to see why I wanted an option: there was a space left blank after the 5 lines of preview so why not fill it with information

by thinking a little more: it is rather the number of lines to import which is interesting, because to take the elevator to see the last line and to estimate the time is not very conclusive.

Finally it is usable as it is, I can import my old users without worries in biblioteQ !

**Reply by textbrowser** [reply]
An idea of a preview is fine, however, now the preview may encompass the entire thing being previewed. So, it's not necessarily a preview. My assumption was that the person importing the data had knowledge of the thing they're importing as well as having the means to review the data elsewhere. Then of course there is this: if the file contains irregularities, another application has to be used. So, the preview is incomplete because another tool is necessary for corrections.
